pygrype/core/backends/docker.py

```diff
@@ -0,0 +1,50 @@
+import subprocess
+from subprocess import CompletedProcess
+
+from pygrype.core.backends.base import GrypeBackend
+
+
+DOCKER_EXE = 'docker'
+IMAGE_NAME = 'anchore/grype'
+
+
+class GrypeDockerBackend(GrypeBackend):
+
+    def __init__(self, tag: str = "latest") -> None:
+        self.tag: str = tag
+        super().__init__()
+
+    @property
+    def docker_image(self) -> str:
+        return f"{IMAGE_NAME}:{self.tag}"
+
+    def ensure_backend(self):
+        try:
+            # Check if Docker is installed
+            subprocess.run(
+                [DOCKER_EXE, '--version'],
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            )
+        except subprocess.CalledProcessError:
+            raise Exception("Docker is not installed or not available in the PATH.")
+
+        try:
+            # Ensure the required Docker image is available
+            subprocess.run(
+                [DOCKER_EXE, 'pull', self.docker_image],
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            )
+        except subprocess.CalledProcessError:
+            raise Exception(f"Failed to pull the required Docker image '{self.docker_image}'.")
+
+    @property
+    def executable_string(self) -> str:
+        return DOCKER_EXE
+
+    def execute(self, *args) -> CompletedProcess:
+        docker_args = ['run', '--rm', self.docker_image] + list(args)
```

Comment:
Creating a container for each command can create unexpected results in scripts where more than one statement is used per grype instance.
```python
grype = Grype(backend=GrypeDockerBackend())
grype.db.update()
for image in ['alpine', 'mariadb', 'ubuntu']:
    grype.scan(image)
```
This script would create 4 separate containers. Reusing the container across all the scans would get a much better performance as the Grype DB is already present after the first pull.

Reply:
Yeah, that would definitely be ideal. However, I'm not sure that it's possible to achieve with the grype docker image as it stands. I've tried creating the container in the `ensure_backend` method of the `GrypeDockerBackend`, and then starting it with `.start()` and `exec()` in the `execute` method. However, the grype dockerfile only defines a single `grype` entrypoint, so attempting to `start()` results in that `grype` executable being run with no arguments, and the container then exits. Subsequent calls to `exec` then fail, as the container is not running.

I'd override that entrypoint to point to a shell, but the Dockerfile uses a Distroless Debian image, so there's no available shell.

I'm not sure of a workaround for this, so I think we're stuck with a container per invocation. If that's the case, i think it make sense to remove them immediately after use so that we're not accumulating exited containers.

Reply:
I thought a little more about this, and I'm pretty sure that I can run the container and set an env var to specify the dB cache location. If I also make that location a /tmp mount on the host, then it should persist beyond the lifetime of the container. It will also mean it's possible to continue to remove the containers after the run.

Reply:
As above, I've updated the pr such that the grype container persists the db in a host mounted tmp directory, which should alleviate some of the performance issues resulting from the container not being able to be long running. Changes are in b2e69cd842b80a204d563bdccb11503abfbe8a78

Reply:
Looks good to me